### ISSUE TITLE
Filesystem resize is required after snapshot restore / volume clone

### ISSUE THREAD

**Issue by jsafrane** [kubernetes/kubernetes#94929]
Kubernetes should resize filesystem on a volume when it's created as a clone of a smaller volume / smaller snapshot is restored into it.

Example:
1. User takes a snapshot of 1 GiB volume with ext3 filesystem on it.
2. User restores the snapshot into 2 GiB volume.
    2.1 CSI driver restores the snapshot into a bigger volume, but ext3 filesystem still uses only 1 GiB of space.
3. CSI driver should make sure that the filesystem on the volume has been resized before the volume is mounted into a pod.

There are few steps needed:

* [x] Clarify behavior of `CreateVolume` CSI call, https://github.com/container-storage-interface/spec/pull/452
* [ ] Update CSI drivers to the new behavior.
  * [x] Move filesystem resize code from kubernetes/kuberenetes to kubernetes/mount-utils, so all CSI drivers can re-use the same code https://github.com/kubernetes/kubernetes/issues/99039 
  * [x] Enhance the code above to compare the volume size from block device perspective and from filesystem perspective. See https://github.com/kubernetes-sigs/aws-ebs-csi-driver/pull/595 for proof of concept. https://github.com/kubernetes/kubernetes/issues/99042
  * [ ] File issues in all community CSI drivers that we care about & fix them.
* [ ] Update csi-tests to (optionally?) check for volume size when cloning / restoring snapshot to a bigger size. https://github.com/kubernetes-csi/csi-test/issues/325

**Comment by humblec** [comment]
True, considering the CSI spec allows restore to a bigger size volume, we should have the support from CO.

**Comment by jsafrane** [comment]
> * Update the external-provisioner to create PVs that are marked for resize when a volume clone / snapshot restore is bigger than the original.

This is hard, Kubernets PV controller overwrites `PVC.Status.Capacity` on binding. 
https://github.com/kubernetes/kubernetes/blob/8422044f1746354393ea5b617e2bde5c24c0f829/pkg/controller/volume/persistentvolume/pv_controller.go#L771-L784

Therefore the external-provisioner cannot create PVC as "marked for resize on node", i.e. with `PVC.Spec.Capacity` = `PV.Spec.Capacity` = the real size of the new volume) and with `PVC.Status.Capacity` = size of the original volume.

We either need to change the binding mechanism not to overwrite `PVC.Status.Size` or add a flag (annotation?) that marks the volume for resize on node, even when `PVC.Spec.Capacity == PVC.Status.Capacity`.

**Comment by jsafrane** [comment]
Talked to @gnufied:

* The external-provisioner could mark `PVC.Status` for resize on node:
  * Set `PVC.Status.Capacity` to the original volume size.
  * Add `FileSystemResizePending` condition to the PVC.
  * The extenal-provisioner needs to modify `PVC.Status`, which needs new RBAC and is a breaking change.
* The PV controller could respect `PVC.Status.Capacity`, if set.

We discussed and rejected adding a new flag / annotation to PV by the external-provisioner for kubelet to call NodeExpandVolume on the first mount. Such a flag would need to be cleared by  kubelet and kubelet can't modify PVs.

**Comment by msau42** [comment]
From latest discussion last week, I believe the conclusion was to have external-provisioner set `PV.Spec.Capacity` to the smaller size. Then we just let the rest of the normal resize flow continue. No need to have provisioner set the conditions.

@kubernetes/sig-storage-bugs

**Comment by gnufied** [comment]
@msau42 we agreed to set conditions too because we want to be as close as possible to external-resizer. If there is some tooling that depends on condition for resizing pending on node, we don't want to break any such tooling. currently kubelet does not depend on condition for node expansion though.

**Comment by msau42** [comment]
Why can't we let resize controller set the condition when it sees pvc size != pv size?

**Comment by gnufied** [comment]
@msau42 hmm good point. I think that should work but we may have to verify that flow.

**Comment by bswartz** [comment]
I think we can do that, but only with a change to the CSI spec wording, because the current implication is that should not be necessary.

**Comment by msau42** [comment]
@jsafrane can you summarize the current conclusion/state of this?

/remove-lifecycle rotten

**Comment by jsafrane** [comment]
Updated the issue with TODO list.

**Comment by xing-yang** [comment]
@jsafrane should this issue stay open to keep track of remaining issues?

**Comment by msau42** [comment]
/triage accepted
/priority long-term
@RomanBednar what is the latest status on this?

**Comment by RomanBednar** [comment]
Hello @msau42, @jsafrane 

I've checked what is the current status of the drivers that were involved in this effort and I think we might be able to close this as all upstream work is done. Please take a look at the table below and let me know if anything is missing.


|Driver |Pull Request |Status |Note|
|-------|-------------|-------|----|
|AWS EBS CSI driver|https://github.com/kubernetes-sigs/aws-ebs-csi-driver/pull/1165|Refactored|Driver used custom code, now it's using mount-utils|
|AWS EFS CSI driver|N/A|N/A|Client resizing not applicable for network filesystems|
|Alibaba CSI driver|https://github.com/kubernetes-sigs/alibaba-cloud-csi-driver/pull/673|Merged|None|
|Azure Disk CSI driver|https://github.com/kubernetes-sigs/azuredisk-csi-driver/pull/1225|Refactored|Driver used custom code due to BTRFS, now it's using mount-utils which was extended to support BTRFS|
|Azure File CSI driver|N/A|Not supported|Restore from snapshot not available in Azure SDK and API: https://github.com/kubernetes-sigs/azurefile-csi-driver/issues/136|
|Cinder CSI driver|https://github.com/kubernetes/cloud-provider-openstack/pull/1563|Merged|None|
|GCP Filestore CSI driver|N/A|N/A|Client resizing not applicable for network filesystems|
|GCP PD CSI driver|https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/pull/972|Merged|e2e test is WIP: https://github.com/kubernetes/kubernetes/pull/110304|
|IBM VPC Block CSI driver|https://github.com/kubernetes-sigs/ibm-vpc-block-csi-driver/pull/100 |Merged|Followup driver fix to align with CSI protocol: https://github.com/kubernetes-sigs/ibm-vpc-block-csi-driver/pull/112|
|oVirt CSI driver|N/A|Not supported|Upstream inactive, downstream not supported (RHV deprecation)|
|vSphere CSI driver|N/A|Not supported|Driver can't restore snap to larger size + clones not supported|
